File: packages/cli/src/catalog.ts
/**
 * Catalog is the phone book. It does not send.
 */
import catalogJson from "../../../vectors/cli_catalog.json" with { type: "json" }

export type CatalogEntry = {
  readonly id: string
  readonly tool: string
  readonly command_tag: string
  readonly summary: string
  readonly channel?: string
  readonly method?: string
}

const CATALOG: ReadonlyArray<CatalogEntry> =
  catalogJson as ReadonlyArray<CatalogEntry>

export const loadCatalog = (): ReadonlyArray<CatalogEntry> => CATALOG

export const getCatalog = (id: string): CatalogEntry => {
  const entry = CATALOG.find((row) => row.id === id)
  if (entry === undefined) {
    throw new Error(`unknown id ${JSON.stringify(id)}; caspian catalog search …`)
  }
  return entry
}

export const searchCatalog = (query: string): ReadonlyArray<CatalogEntry> => {
  const words = query
    .split(/\s+/)
    .map((word) => word.toLowerCase())
    .filter((word) => word.length > 2)
  if (words.length === 0) return loadCatalog()
  return CATALOG.filter((entry) => {
    const hay = Object.values(entry).join(" ").toLowerCase()
    return words.every((word) => hay.includes(word))
  })
}

File: apps/cli/src/caspian_cli/catalog.py
"""Catalog is the phone book. It does not send."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

_CACHE: list[dict[str, Any]] | None = None


def _catalog_path() -> Path:
    here = Path(__file__).resolve()
    for parent in here.parents:
        candidate = parent / "vectors" / "cli_catalog.json"
        if candidate.is_file():
            return candidate
    raise FileNotFoundError("vectors/cli_catalog.json not found")


def load_catalog() -> list[dict[str, Any]]:
    global _CACHE
    if _CACHE is None:
        _CACHE = json.loads(_catalog_path().read_text())
    return list(_CACHE)


def get_catalog(id: str) -> dict[str, Any]:
    for entry in load_catalog():
        if entry["id"] == id:
            return entry
    raise KeyError(id)


def search_catalog(query: str) -> list[dict[str, Any]]:
    words = [w.lower() for w in query.split() if len(w) > 2]
    if not words:
        return load_catalog()
    hits: list[dict[str, Any]] = []
    for entry in load_catalog():
        hay = " ".join(str(v) for v in entry.values()).lower()
        if all(word in hay for word in words):
            hits.append(entry)
    return hits
